### PR TITLE
Add 3DS example VC that uses PaymentIntents to charge the user

### PR DESCRIPTION
**Pull request by danj-stripe**
## Summary
Add 3DS example VC that uses PaymentIntents to charge the user

This is built on top of the existing 3DS example VC (using the same UI),
but intercepts the `pay` button response to use a PaymentIntent instead.

This relies on #986 and #987.

## Motivation
IOS-792

## Testing
Ran the example VC and played around with it. This was used as the test app for the previous
PRs.

### CODE REVIEW

**Comment by danj-stripe** [comment]
Since this PR relies on several previous ones, they're all merged into here. The changes specific to this PR start at https://github.com/stripe/stripe-ios/pull/988/commits/33fd12c8f5f9f77a634621be6c17d2dabd780a77

File: Example/Custom Integration (ObjC)/BrowseExamplesViewController.m [hunk 1]
```diff
@@ -36,24 +37,27 @@ - (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
...
+        cell.textLabel.text = @"Apple Pay";
+        break;
         case 3:
-            cell.textLabel.text = @"Sofort";
-            break;
+        cell.textLabel.text = @"Sofort";
+        break;
+        case 4:
+        cell.textLabel.text = @"Card + 3DS + PaymentIntent";
+        break;
```
Comment (csabol-stripe) [inline Example/Custom Integration (ObjC)/BrowseExamplesViewController.m]: does `ctrl+i` give this indentation for obj-c now?

File: Example/Custom Integration (ObjC)/BrowseExamplesViewController.m [hunk 2]
```diff
@@ -123,7 +133,69 @@ - (void)createBackendChargeWithSource:(NSString *)sourceID completion:(STPSource
                                                           if (error) {
                                                               completion(STPBackendResultFailure, error);
                                                           } else {
-                                                              completion(STPBackendChargeResultSuccess, nil);
+                                                              completion(STPBackendResultSuccess, nil);
+                                                          }
+                                                      }];
+
+    [uploadTask resume];
```
Comment (csabol-stripe) [inline Example/Custom Integration (ObjC)/BrowseExamplesViewController.m]: why was this added?

File: Example/Custom Integration (ObjC)/BrowseExamplesViewController.m [hunk 3]
```diff
@@ -123,7 +133,69 @@ - (void)createBackendChargeWithSource:(NSString *)sourceID completion:(STPSource
...
+}
+
+
+/**
+ Ask the example backed to create a PaymentIntent with the specified amount.
```
Comment (csabol-stripe) [inline Example/Custom Integration (ObjC)/BrowseExamplesViewController.m]: typo, `backend`

File: Example/Custom Integration (ObjC)/BrowseExamplesViewController.m [hunk 4]
```diff
@@ -123,7 +133,69 @@ - (void)createBackendChargeWithSource:(NSString *)sourceID completion:(STPSource
...
+
+ The implementation of this function is not interesting nor relevant to using PaymentIntents. The
```
Comment (csabol-stripe) [inline Example/Custom Integration (ObjC)/BrowseExamplesViewController.m]: grammar nit: `not interesting or relevant`

Comment (csabol-stripe) [reply Example/Custom Integration (ObjC)/BrowseExamplesViewController.m]: https://www.quickanddirtytips.com/education/grammar/when-use-nor?page=1

File: Example/Custom Integration (ObjC)/ThreeDSPaymentIntentExampleViewController.m [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+                [self.delegate exampleViewController:self didFinishWithError:error];
+                return;
+            }
+
+            if (paymentIntent.status == STPPaymentIntentStatusRequiresSourceAction) {
+                self.redirectContext = [[STPRedirectContext alloc] initWithPaymentIntent:paymentIntent returnUrl:returnUrl completion:^(NSString * _Nonnull clientSecret, NSError * _Nullable error) {
+                    if (error) {
+                        [self.delegate exampleViewController:self didFinishWithError:error];
+                    }
+                    else {
```
Comment (csabol-stripe) [inline Example/Custom Integration (ObjC)/ThreeDSPaymentIntentExampleViewController.m]: this file is mixing `} else {` and `}\n else {` Do we have a standard in our sample code? I'd vote for `} else {`

File: Example/Custom Integration (ObjC)/ThreeDSPaymentIntentExampleViewController.m [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+                        [stripeClient retrievePaymentIntentWithClientSecret:clientSecret completion:^(STPPaymentIntent * _Nullable paymentIntent, NSError * _Nullable error) {
+                            if (error) {
+                                [self.delegate exampleViewController:self didFinishWithError:error];
+                            } else {
+                                [self finishWithStatus:paymentIntent.status];
+                            }
+                        }];
+                    }
+                    self.redirectContext = nil; // break retain cycle
```
Comment (csabol-stripe) [inline Example/Custom Integration (ObjC)/ThreeDSPaymentIntentExampleViewController.m]: so it's an intentional retain cycle? Can we call that out in like L72?

Comment (csabol-stripe) [reply Example/Custom Integration (ObjC)/ThreeDSPaymentIntentExampleViewController.m]: Is there a not retain cycle way to accomplish this? Doesn't feel like we should be demonstrating to users that a retain cycle is the "right" way to do something

File: Example/Custom Integration (ObjC)/ThreeDSPaymentIntentExampleViewController.m (hunk 5 above)
Comment (joeydong-stripe) [reply Example/Custom Integration (ObjC)/ThreeDSPaymentIntentExampleViewController.m]: We encoded `}\n else {` in our style guide a long time ago but I'm coming around to doing `} else {` everywhere. Should open a PR that fixes this across the codebase.

Comment (joeydong-stripe) [reply Example/Custom Integration (ObjC)/ThreeDSPaymentIntentExampleViewController.m]: https://github.com/stripe/stripe-ios/blob/master/STYLEGUIDE.md#control-flow

File: Example/Custom Integration (ObjC)/BrowseExamplesViewController.m (hunk 2 above)
Comment (danj-stripe) [reply Example/Custom Integration (ObjC)/BrowseExamplesViewController.m]: This is pre-existing code, it's just the Github diff algorithm is being weird. Both methods end the same way (L202 has the same call).

File: Example/Custom Integration (ObjC)/ThreeDSPaymentIntentExampleViewController.m (hunk 5 above)
Comment (danj-stripe) [reply Example/Custom Integration (ObjC)/ThreeDSPaymentIntentExampleViewController.m]: I don't think I've ever seen `}\n else {` in a style guide, and it wouldn't be my choice, but I try to adhere to it. Thanks for catching the mistake.

File: Example/Custom Integration (ObjC)/ThreeDSPaymentIntentExampleViewController.m (hunk 6 above)
Comment (danj-stripe) [reply Example/Custom Integration (ObjC)/ThreeDSPaymentIntentExampleViewController.m]: I think it starts on L79 - `self` needs to retain the `redirectContext` (doesn't work otherwise), and none of the examples are using weak self, so the `redirectContext` retains `self`.

`SofortExampleViewController` and `ThreeDSExampleViewController` do the same thing, but without this note.

It's possible that it's also necessary so that `STPRedirectContext.dismissPresentedViewController` is called, and it might be intentional that we're releasing the memory early.

IMO our example code does not do a great job demonstrating best practices, instead preferring to do things compactly with very little indirection, so that "all the pieces" are in one place.

I think there are improvements that could be made, and I think there's value in keeping the examples in sync.

File: Example/Custom Integration (ObjC)/BrowseExamplesViewController.m (hunk 1 above)
Comment (danj-stripe) [reply Example/Custom Integration (ObjC)/BrowseExamplesViewController.m]: It was, and I didn't like it but couldn't get it to do anything different. Now it looks like Xcode has changed its mind. So there must have been something I'd missed, maybe an error in a header or something.

**Comment by danj-stripe** [comment]
In order to fix conflicts and pull in latest changes from #986 and #987, rebased onto master, merged in latest from those PRs, and then picked the 4 commits for this PR. Starting after the octopus merge (6b79bb7) has the changes to the example code, and they're the same as reviewed by csabol, plus f035539 with updates

File: Example/Custom Integration (ObjC)/ThreeDSPaymentIntentExampleViewController.m (hunk 6 above)
Comment (csabol-stripe) [reply Example/Custom Integration (ObjC)/ThreeDSPaymentIntentExampleViewController.m]: makes sense. Thanks

**Comment by danj-stripe** [comment]
This now includes #994 and #995, and has been rebased onto master. No new or un-reviewed code
